{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "878b5946",
   "metadata": {},
   "source": [
    "## omibio Analysis demo\n",
    "This is an example of using omibio for process analysis. We used omibio to analyze a sequence of the *E. coli* genome. The main contents include:\n",
    "- Read fasta files.\n",
    "- Predict Open Reading Frames (ORFs).\n",
    "- Write the prediction ORF as a FASTA file.\n",
    "- Perform sliding window GC content calculation.\n",
    "- Perform motif searches.\n",
    "- Visualize ORFs, sliding window GC, and motifs.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b71b738f",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "find_orfs() argument 'seq_id' must be str, not Sequence",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 20\u001b[39m\n\u001b[32m     17\u001b[39m sequence: Sequence = seq_entry.seq\n\u001b[32m     19\u001b[39m \u001b[38;5;66;03m# Find ORFs\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m20\u001b[39m orf_result: AnalysisResult = \u001b[43mfind_orfs\u001b[49m\u001b[43m(\u001b[49m\u001b[43msequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmin_length\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m150\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minclude_reverse\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mseq_id\u001b[49m\u001b[43m=\u001b[49m\u001b[43mseq_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     22\u001b[39m \u001b[38;5;66;03m# Sliding GC content calculation\u001b[39;00m\n\u001b[32m     23\u001b[39m gc_result: AnalysisResult = sliding_gc(sequence, seq_id=seq_id, window=\u001b[32m50\u001b[39m, step=\u001b[32m10\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mD:\\python_projects\\omiBioKit\\omibio\\analysis\\find_orfs.py:152\u001b[39m, in \u001b[36mfind_orfs\u001b[39m\u001b[34m(seq, min_length, max_length, overlap, include_reverse, sort_by_length, translate, start_codons, seq_id)\u001b[39m\n\u001b[32m    150\u001b[39m \u001b[38;5;66;03m# Input: seq_id\u001b[39;00m\n\u001b[32m    151\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m seq_id \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(seq_id, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m--> \u001b[39m\u001b[32m152\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[32m    153\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mfind_orfs() argument \u001b[39m\u001b[33m'\u001b[39m\u001b[33mseq_id\u001b[39m\u001b[33m'\u001b[39m\u001b[33m must be str, not \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    154\u001b[39m         + \u001b[38;5;28mtype\u001b[39m(seq_id).\u001b[34m__name__\u001b[39m\n\u001b[32m    155\u001b[39m     )\n\u001b[32m    157\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(seq, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m    158\u001b[39m     seq = Sequence(seq, strict=\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[31mTypeError\u001b[39m: find_orfs() argument 'seq_id' must be str, not Sequence"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "from omibio import *\n",
    "from omibio.io import read_fasta, write_fasta\n",
    "from omibio.analysis import find_orfs, sliding_gc, find_motifs\n",
    "import re\n",
    "\n",
    "nb_dir = Path().resolve()\n",
    "\n",
    "input_fasta_path = nb_dir.parent / \"data\" / \"e_coli_gene_sequence.fasta\"\n",
    "output_fasta_path = nb_dir.parent / \"output\" / \"demo_output.fasta\"\n",
    "\n",
    "\n",
    "# Get sequence id & sequence\n",
    "seqs: SeqCollections = read_fasta(input_fasta_path)\n",
    "seq_entry: SeqEntry = seqs.entry_list()[0]\n",
    "sequence: Sequence = seq_entry.seq\n",
    "\n",
    "# Find ORFs\n",
    "orf_result: AnalysisResult = find_orfs(sequence, min_length=150, include_reverse=True)\n",
    "\n",
    "# Sliding GC content calculation\n",
    "gc_result: AnalysisResult = sliding_gc(sequence, window=50, step=10)\n",
    "\n",
    "# Motifs\n",
    "pattern = re.compile(r'(AGGAGG|AGGAG|GGAGG|AGGA|GAGG|GGAG).{5,12}ATG')\n",
    "motif_result: AnalysisResult = find_motifs(sequence, pattern, include_reverse=True)\n",
    "\n",
    "# Visualization & output results\n",
    "orf_dict: dict= orf_result.to_dict(prefix=\"EcoliORFs\")\n",
    "lines = write_fasta(output_fasta_path, orf_dict, line_len=80)\n",
    "for line in lines:\n",
    "    print(line)\n",
    "\n",
    "orf_result.plot()\n",
    "gc_result.plot()\n",
    "motif_result.plot()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80112764",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "In this analysis, we read a DNA sequence from E. coli and performed three types of analysis: ORF lookup, sliding GC calculation, and motif lookup.\n",
    "\n",
    "- First, we conduct a preliminary analysis of the sequence.\n",
    "    - For ORF lookup, we set the minimum ORF length to 150bp `(min_length=150)` to filter out shorter “false positive” ORFs; we also look for the reverse complementary sequence of the sequence `(include_reverse=True)` to obtain the maximum probability.\n",
    "    - For the calculation of sliding GC, we set the window length to a small 50bp and the step size to 10bp `(window=50, step=10)`. The reason for this is that the overall DNA sequence GC fluctuations in eukaryotes are relatively insignificant, and a smaller window length can more sensitively capture small GC peaks.\n",
    "    - For motif searching, we used regular expression searching and compiled the following pattern: `re.compile(r'(AGGAGG|AGGAG|GGAGG|AGGA|GAGG|GGAG).{5,12}ATG')`. In this regular expression, the last `'ATG'` is the start codon of the encoded sequence, while the preceding A and G sequences are several common ribosome binding site (RBS) sequences. They typically appear upstream of the start codon.`'{5,12}'` then defines the distance between these two sequences. Similarly, we also search for motif in reverse complementary sequences`(include_reverse=True)`.\n",
    "\n",
    "- Then, we output and visualize the results.\n",
    "    - We can use the `write_fasta()` function to write ORF results into a fasta file.\n",
    "    - we can perform simple visualization of the analysis results by using the `.plot()` method of `AnalysisResult` (returns of `find_orfs()`, `sliding_gc()`, `find_motifs()`).\n",
    "\n",
    "- Finally, we analyzed and compared the results.\n",
    "    - From the first ORF distribution chart, we found a relatively long ORF in the +1 frame of the positive chain.\n",
    "    - From the second sliding GC chart, we found that the sequence has several GC peaks, which are located roughly within the long ORF we previously found.\n",
    "    - In the third motif distribution chart, we found a motif matching the pattern before the previously determined start position of the long ORF. This is very strong evidence.\n",
    "\n",
    "Based on the above analysis, we can roughly determine that this DNA sequence contains a gene. We can conduct further analysis, such as determining the specific location of the gene based on the `start` and `end` attributes of the `SeqInterval` class stored in `AnalysisResult` from `find_orfs`. We won't go into details here.\n",
    "In fact, the gene we found in this analysis is the gene encoding the shikimate transport protein (shiA) that actually exists in E. coli. Hopefully, this use case will give you an understanding of the features and some usage methods of omibio.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
